models/src/node/mod.rs

```diff
@@ -0,0 +1,259 @@
+/// and the host agent drives state changes forward and updates the `BottlerocketNodeStatus`.
+#[derive(
+    Clone,
+    CustomResource,
+    Serialize,
+    Deserialize,
+    Debug,
+    Default,
+    Eq,
+    PartialEq,
+    JsonSchema,
+    Validate,
+)]
+#[kube(
+    derive = "Default",
+    derive = "PartialEq",
+    group = "brupop.bottlerocket.aws",
+    kind = "BottlerocketNode",
+    namespaced,
+    plural = "bottlerocketnodes",
+    shortname = "brn",
+    singular = "bottlerocketnode",
+    status = "BottlerocketNodeStatus",
+    version = "v1",
+    printcolumn = r#"{"name":"State", "type":"string", "jsonPath":".status.current_state"}"#,
+    printcolumn = r#"{"name":"Version", "type":"string", "jsonPath":".status.current_version"}"#,
+    printcolumn = r#"{"name":"Target State", "type":"string", "jsonPath":".spec.state"}"#,
+    printcolumn = r#"{"name":"Target Version", "type":"string", "jsonPath":".spec.version"}"#
+)]
+pub struct BottlerocketNodeSpec {
+    /// Records the desired state of the `BottlerocketNode`
+    pub state: BottlerocketNodeState,
+    /// The time at which the most recent state was set as the desired state.
+    state_transition_timestamp: Option<String>,
+    /// The desired update version, if any.
+    #[validate(regex = "SEMVER_RE")]
+    version: Option<String>,
+}
+
+impl BottlerocketNode {
+    /// Creates a `BottlerocketNodeSelector` from this `BottlerocketNode`.
+    pub fn selector(&self) -> Result<BottlerocketNodeSelector> {
+        BottlerocketNodeSelector::from_bottlerocket_node(self)
+    }
+
+    /// Returns whether or not a node has reached the state requested by its spec.
+    pub fn has_reached_desired_state(&self) -> bool {
+        self.status.as_ref().map_or(false, |node_status| {
+            node_status.current_state == self.spec.state
+        })
+    }
+}
+
+impl BottlerocketNodeSpec {
+    pub fn new(
+        state: BottlerocketNodeState,
+        state_transition_timestamp: Option<DateTime<Utc>>,
+        version: Option<Version>,
+    ) -> Self {
+        let state_transition_timestamp = state_transition_timestamp.map(|ts| ts.to_rfc3339());
+        let version = version.map(|v| v.to_string());
+        BottlerocketNodeSpec {
+            state,
+            state_transition_timestamp,
+            version,
+        }
+    }
+
+    /// Creates a new BottlerocketNodeSpec, using the current time as the timestamp for timeout purposes.
+    pub fn new_starting_now(state: BottlerocketNodeState, version: Option<Version>) -> Self {
+        Self::new(state, Some(Utc::now()), version)
+    }
+
+    /// JsonSchema cannot appropriately handle DateTime objects. This accessor returns the transition timestamp
+    /// as a DateTime.
+    pub fn state_timestamp(&self) -> Option<DateTime<Utc>> {
+        self.state_transition_timestamp.as_ref().map(|ts_str| {
+            DateTime::parse_from_rfc3339(ts_str)
+                .expect("state_transition_timestamp must be rfc3339 string.")
```

Comment:
Shouldn't this use `context`?

Reply:
I had the debate about writing this on whether or not this indicated a software bug or an actual error and have decided that even in either scenario, returning an `Err` value is probably best. I'll change it.